README.md
# ![Rusoto](./assets/logo-wide.png)

<a href="https://github.com/rusoto/rusoto/actions?query=workflow%3A%22Build+and+test%22"><img src="https://github.com/rusoto/rusoto/workflows/Build%20and%20test/badge.svg"></a>
<a href="https://docs.rs/rusoto_core" title="API Docs"><img src="https://img.shields.io/badge/API-docs-blue.svg" alt="api-docs-badge"></img></a>
<a href="https://crates.io/crates/rusoto_core" title="Crates.io"><img src="https://img.shields.io/crates/v/rusoto_core.svg" alt="crates-io"></img></a>
<a href="#license" title="License: MIT"><img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="license-badge"></img></a>
<a href="https://deps.rs/repo/github/rusoto/rusoto" title="Dependency Status"><img src="https://deps.rs/repo/github/rusoto/rusoto/status.svg" alt="dependency-status-badge"></img></a>
<a href="https://discordapp.com/invite/WMJ4DWp"><img src="https://img.shields.io/discord/670751965273391124"></img></a>

**Rusoto is an AWS SDK for Rust**

---

You may be looking for:

* [An overview of Rusoto][rusoto-overview]
* [AWS services supported by Rusoto][supported-aws-services]
* [API documentation][api-documentation]
* [Getting help with Rusoto][rusoto-help]

## Maintenance status

⚠️ Rusoto is in **maintenance mode.** ⚠️

The current maintainers only have the bandwidth to review dependency bumps and
obvious bugfixes. Our bandwidth for reviewing new features is extremely limited.

While you are welcome to submit PRs that implement new features or refactor
existing code, they are unlikely to be merged unless we can find more active
maintainers.

Please see [Meta: future of Rusoto (#1651)](https://github.com/rusoto/rusoto/issues/1651)
for details.

## Installation

Rusoto is available on [crates.io](https://crates.io/crates/rusoto_core).
To use Rusoto in your Rust program built with Cargo, add it as a dependency and `rusoto_$SERVICENAME` for any supported AWS service you want to use.

For example, to include only S3 and SQS:

```toml
[dependencies]
rusoto_core = "0.48.0"
rusoto_sqs = "0.48.0"
rusoto_s3 = "0.48.0"
```

## Migration notes

Breaking changes and migration details are documented at [https://rusoto.org/migrations.html](https://rusoto.org/migrations.html).

Note that from v0.43.0 onward, Rusoto uses Rust's `std::future::Future`, and the Tokio 0.2 ecosystem. From v0.46.0 onward, Rusoto uses the Tokio 1.0 ecosystem.

## Usage

Rusoto has a crate for each AWS service, containing Rust types for that service's API.
A full list of these services can be found [here][supported-aws-services].
All other public types are reexported to the crate root.
Consult the rustdoc documentation for full details by running `cargo doc` or visiting the online [documentation](https://docs.rs/rusoto_core) for the latest crates.io release.

A simple example of using Rusoto's DynamoDB API to list the names of all tables in a database:

```rust,no_run
use rusoto_core::Region;
use rusoto_dynamodb::{DynamoDb, DynamoDbClient, ListTablesInput};

#[tokio::main]
async fn main() {
    let client = DynamoDbClient::new(Region::UsEast1);
    let list_tables_input: ListTablesInput = Default::default();

    match client.list_tables(list_tables_input).await {
        Ok(output) => match output.table_names {
            Some(table_name_list) => {
                println!("Tables in database:");

                for table_name in table_name_list {
                    println!("{}", table_name);
                }
            }
            None => println!("No tables in database!"),
        },
        Err(error) => {
            println!("Error: {:?}", error);
        }
    }
}
```

### Credentials

For more information on Rusoto's use of AWS credentials such as priority and refreshing, see [AWS Credentials](AWS-CREDENTIALS.md).

## Semantic versioning

Rusoto complies with [semantic versioning 2.0.0](http://semver.org/).
Until reaching 1.0.0 the API is to be considered unstable.
See [Cargo.toml](Cargo.toml) or [rusoto on crates.io](https://crates.io/crates/rusoto_core) for current version.

## Releases

Information on release schedules and procedures are in [RELEASING](RELEASING.md).

## Contributing

Discussions take place on the [Rusoto Discord channel][discord-invite].

See [CONTRIBUTING](CONTRIBUTING.md) for more information.

## Supported OSs, Rust versions and non-AWS projects

Linux, macOS and Windows are supported and tested via [GitHub actions](https://github.com/rusoto/rusoto/actions?query=workflow%3A%22Build+and+test%22).

Rust stable, beta and nightly are supported.

Rusoto's primary aim is to be used with AWS. Other projects that provide AWS-like APIs, such as Ceph, Minio, Yandex Object Storage, etc... are not a focus at this time. PRs to fix issues with Rusoto and AWS-like APIs are welcome but generally won't be created by Rusoto maintainers.

## License

Rusoto is distributed under the terms of the MIT license.

See [LICENSE][license] for details.

[api-documentation]: https://docs.rs/rusoto_core "API documentation"
[license]: https://github.com/rusoto/rusoto/blob/master/LICENSE "MIT License"
[rusoto-help]: https://www.rusoto.org/help.html "Getting help with Rusoto"
[rusoto-overview]: https://www.rusoto.org/ "Rusoto overview"
[supported-aws-services]: https://www.rusoto.org/supported-aws-services.html "List of AWS services supported by Rusoto"
[discord-invite]: https://discordapp.com/invite/WMJ4DWp "Discord channel invite"


CHANGELOG.md
# Rusoto changes

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

(Please put changes here.)

## [0.48.0] - 2022-04-24

- Remove [`macie`](https://github.com/boto/botocore/blob/81e0b0026f9eb21be658cf0dd4c17e0dc173346f/botocore/data/macie/2017-12-19/service-2.json#L27), [`mobile`](https://stackoverflow.com/a/64990655), and [`worklink`](https://github.com/boto/botocore/blob/81e0b0026f9eb21be658cf0dd4c17e0dc173346f/botocore/data/worklink/2018-09-25/service-2.json#L33-L34) services, which have had their DNS records removed
- Add support for ap-southeast-3, Asia Pacific (Jakarta)
- Add Clone derived trait to TlsError
- Allow the user to choose between rustls and rustls-webpki, and make only the former depend on native cert support
- Update `hyper-rustls` to `0.23`
- Avoid caching `CredentialError` in `AutoRefreshingProvider`.

## [0.47.0] - 2021-06-29

- Update to botocore 1.20.102
- Update to `serde_urlencoded` 0.7
- Update to `rustc_version` 0.4
- Update to `shlex` 1.0
- Update to  `hmac` 0.4
- Replace `time`-related types in `rusoto_signature` with `chrono` types, to
  match `rusoto_credential`
- Swap the non-RustCrypto `md5` crate for the RustCrypto `md-5` crate, to match
  usage of RustCrypto `sha2` crate
- Remove `Sync` constraint on `ByteStream`-related functions.
- Update incorrect minimum versions of `async-trait`, `percent-encoding`, and
  `serde`
- Avoid unnecessary calls to `to_string`
- Remove unpublished and outdated `helpers` crate

## [0.46.0] - 2021-01-05

- Display `rusoto_core::Client` in docs
- Fix unsoundness in `rusoto_mock::MultipleMockRequestDispatcher`
- Add ability to set local agent appended to the default User-Agent
- Update to `base64` 0.13
- Update to `bytes` 1.0
- Update to `hmac` 0.10
- Update to `hyper-rustls` 0.22
- Update to `hyper-tls` 0.5
- Update to `hyper` 0.14
- Update to `tokio` 1.0
- Update to botocore 1.19.42
- Swap the unmaintained `dirs` crate for its replacement `dirs-next`
- Swap `pin-project` for the lighter weight `pin-project-lite`
- Disable `chrono`'s `oldtime` feature
- Remove dependency on `regex`

## [0.45.0] - 2020-07-22

- Add event-stream protocol support (currently only for JSON APIs, used in `subscribe_to_shard` call in Kinesis)
- Extract common generated code into utility functions to improve compile times
- Allow creating a ProfileProvider with only the profile
- CDATA sections are now treated like strings
- Fix incorrect type definition for `rusoto_batch::JobDetail`
- Update to `hmac` 0.8 and `sha2` 0.9
- Added `Sync` bounds to `AsyncRead` and `Read` structures
- Update to botocore 1.17.20

## [0.44.0] - 2020-06-01

- Add support for af-south-1, Africa (Cape Town), and eu-south-1, Europe (Milan)
- Update to botocore 1.16.14
- Fix `Time::now()` and `OffsetDateTime::now()` deprecation warnings
- Fix minimum version of `time` crate
- Always encode `+` in query strings
- Added a Cognito credential provider
- Add `MultipleMockRequestDispatcher` to permit mocking multiple requests using the same client
- Fix `rusoto_sts::WebIdentityProvider::from_k8s_env` always requiring `AWS_ROLE_SESSION_NAME` env var which should be optional
- Added support to optionally define a session policy when using `rusoto_sts::WebIdentityProvider`
- Omit generating XML-deseralization code for actions without a response body
- Add `region_from_profile()` function to ProfileProvider
- Fix applying `Content-Encoding`
- Added `new_with_size()` function to ByteStream
- Add defualt help text to Makefile

## [0.43.0] - 2020-03-15

- Fix minimum version of hyper
- Fix `PrimitiveDateTime` deprecation error
- Update to dirs 2.0
- Bump base64 to 0.12 and hyper-rustls to 0.20
- Fix `serialize_structs` and `deserialize_structs`
- Fix JWT serialization in `WebIdentityProvider`
- Add ability to set local agent prepended to the default User-Agent
- Fix invalid signature for Route 53 `resource_record_sets` methods
- Improve `Display` impl for `RusotoError::Unknown`
- Fix hang in XML deserialization for flattened shapes
- Remove obsolete `RusotoFuture` and fix docs generation on nightly
- Fix `credential_process`, again
- Change non-China S3 domains to `s3.{region}.amazonaws.com`

## [0.43.0-beta.1] - 2020-02-07

- Move to `std::future::Future`, async/.await, and Tokio 0.2
- Update to botocore 1.14.9
- Add Discord invite link to README.md and CONTRIBUTING.md
- Remove unused import
- Fixed links in AWS-CREDENTIALS.md
- Fixed SNS API's attributes and value keyword
- Adding support for web identity provider, which enables IAM roles for Kubernetes service accounts.
- Add object-safe AwsCredentialsProvider trait as alternative to the existing generic `ProvideAwsCredentials` trait.
- Introduce `Secret` type to automatically zero-out memory use to stored secret credentials. So far, only used in the new web identity provider.
- Introduce `Variable` to abstract over certain credential provider input parameters.
- Encode request payload optionally with Gzip
- Add Debug trait to generated Clients
- Add `rusoto_ec2::filter!` macro
- Improve `InstanceMetadataProvider` to avoid cloning unnecessarily
- Remove deprecated `Error::description` implementations
- Add features `serialize_structs` and `deserialize_structs`
- Implement Clone on various Credential structs.
- Fix incorrect encoding of Session Token when pre-signing URLs
- Add IoT Secure Tunneling service
- Fix Directory Service integration tests
- Update to time 0.2.x

## [0.42.0] - 2019-11-18

- Use static initializer for AWS profile regex
- Add QLDB service
- Add QLDB Session service
- Update Skeptic tests for Rusoto v0.41
- Don't decode query string parameters before encoding it. Results in fixing the prefix and marker
params for s3 `list_objects` methods
- Add Textract service
- Update CloudDirectory API definition to `2017-01-11`
- Add SecurityHub service
- Add Transfer service

build.rs
extern crate rusoto_codegen;
extern crate rayon;

use std::env;
use std::path::Path;

use rusoto_codegen::{Service, generate};
use rayon::prelude::*;

/*
gamelift/2015-10-01/service-2.json:    "protocol":"json"
support/2013-04-15/service-2.json:    "protocol":"json"
*/

// expand to use cfg!() so codegen only gets run for services
// in the features list
macro_rules! services {
    ( $( [$name:expr, $date:expr] ),* ) => {
        {
            let mut services = Vec::new();
            $(
                if cfg!(feature = $name) {
                    services.push(Service::new($name, $date));
                }
            )*
            services
        }
    }
}

fn main() {
    let out_dir = env::var_os("OUT_DIR").expect("OUT_DIR not specified");
    let out_path = Path::new(&out_dir).to_owned();

    let services = services! {
        ["acm", "2015-12-08"],
        ["autoscaling", "2011-01-01"],
        ["cloudformation", "2010-05-15"],
        ["cloudfront", "2016-11-25"],
        ["cloudhsm", "2014-05-30"],
        ["cloudsearch", "2013-01-01"],
        ["cloudtrail", "2013-11-01"],
        ["cloudwatch", "2010-08-01"],
        ["codecommit", "2015-04-13"],
        ["codedeploy", "2014-10-06"],
        ["codepipeline", "2015-07-09"],
        ["cognito-identity", "2014-06-30"],
        ["config", "2014-11-12"],
        ["datapipeline", "2012-10-29"],
        ["devicefarm", "2015-06-23"],
        ["directconnect", "2012-10-25"],
        ["ds", "2015-04-16"],
        ["dynamodb", "2012-08-10"],
        ["dynamodbstreams", "2012-08-10"],
        ["ec2", "2016-11-15"],
        ["ecr", "2015-09-21"],
        ["ecs", "2014-11-13"],
        ["elasticache", "2015-02-02"],
        ["elasticbeanstalk", "2010-12-01"],
        ["elastictranscoder", "2012-09-25"],
        ["elb", "2012-06-01"],
        ["elbv2", "2015-12-01"],
        ["emr", "2009-03-31"],
        ["events", "2015-10-07"],
        ["firehose", "2015-08-04"],
        ["iam", "2010-05-08"],
        ["importexport", "2010-06-01"],
        ["inspector", "2016-02-16"],
        ["iot", "2015-05-28"],
        ["kinesis", "2013-12-02"],
        ["kms", "2014-11-01"],
        ["lambda", "2015-03-31"],
        ["logs", "2014-03-28"],
        ["machinelearning", "2014-12-12"],
        ["marketplacecommerceanalytics", "2015-07-01"],
        ["opsworks", "2013-02-18"],
        ["redshift", "2012-12-01"],
        ["rds", "2014-10-31"],
        ["route53", "2013-04-01"],
        ["route53domains", "2014-05-15"],
        ["s3", "2006-03-01"],
        ["sdb", "2009-04-15"],
        ["ses", "2010-12-01"],
        ["sns", "2010-03-31"],
        ["sqs", "2012-11-05"],
        ["ssm", "2014-11-06"],
        ["storagegateway", "2013-06-30"],
        ["sts", "2011-06-15"],
        ["swf", "2012-01-25"],
        ["waf", "2015-08-24"],
        ["workspaces", "2015-04-08"],
        ["xray", "2016-04-12"]
    };

    let count: usize =
        services.into_par_iter().map(|service| generate(service, &out_path.clone())).count();
    println!("\nGenerated {:?} services.\n", count);

    let codegen_dir = Path::new("codegen");

    // avoid unnecessary recompiles when used as a crates.io dependency
    if codegen_dir.exists() {
        println!("cargo:rerun-if-changed=codegen");
    }
}
